File: tests/CodexQuotaRail.Core.Tests/QuotaNormalizerTests.cs
using CodexQuotaRail.Core.Quotas;

namespace CodexQuotaRail.Core.Tests;

public sealed class QuotaNormalizerTests
{
    [Theory]
    [InlineData(0, 100)]
    [InlineData(68, 32)]
    [InlineData(100, 0)]
    [InlineData(-5, 100)]
    [InlineData(120, 0)]
    public void ConvertsUsedToAvailable(int used, int expected)
    {
        var source = new RawQuotaWindow("5 小时", used, 300, 1_800_000_000, false);
        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(expected, result.AvailablePercent);
    }

    [Fact]
    public void ExtremeNegativeUsedPercentIsClampedBeforeConverting()
    {
        var source = new RawQuotaWindow("5 小时", int.MinValue, 300, null, false);

        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(100, result.AvailablePercent);
        Assert.Equal(QuotaWindowState.Healthy, result.State);
    }

    [Fact]
    public void MissingUsedPercentIsUnavailableNotFull()
    {
        var source = new RawQuotaWindow("5 小时", null, 300, null, false);
        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(QuotaWindowState.Unavailable, result.State);
        Assert.Null(result.AvailablePercent);
    }

    [Fact]
    public void OmitsMissingSecondaryWindow()
    {
        var source = new RawQuotaSnapshot(
            new RawQuotaWindow("5 小时", 40, 300, null, false),
            null,
            DateTimeOffset.UnixEpoch);

        var result = QuotaNormalizer.Normalize(source);

        Assert.Single(result.Windows);
    }

    [Fact]
    public void WaitingCreatesConnectingStateWithoutQuotaWindows()
    {
        var result = QuotaDisplayState.Waiting("正在连接 Codex");

        Assert.Empty(result.Windows);
        Assert.Equal(QuotaConnectionState.Connecting, result.Connection);
        Assert.Null(result.UpdatedAt);
        Assert.Equal("正在连接 Codex", result.Message);
    }

    [Fact]
    public void NormalizesBothWindowsInSourceOrderAndMarksSnapshotLive()
    {
        var receivedAt = DateTimeOffset.FromUnixTimeSeconds(1_800_000_100);
        var source = new RawQuotaSnapshot(
            new RawQuotaWindow("5 小时", 40, 300, null, false),
            new RawQuotaWindow("本周", 20, 10_080, null, false),
            receivedAt);

        var result = QuotaNormalizer.Normalize(source);

        Assert.Collection(
            result.Windows,
            primary => Assert.Equal("5 小时", primary.Label),
            secondary => Assert.Equal("本周", secondary.Label));
        Assert.Equal(QuotaConnectionState.Live, result.Connection);
        Assert.Equal(receivedAt, result.UpdatedAt);
        Assert.Null(result.Message);
    }

    [Theory]
    [InlineData(49, QuotaWindowState.Healthy)]
    [InlineData(50, QuotaWindowState.Notice)]
    [InlineData(79, QuotaWindowState.Notice)]
    [InlineData(80, QuotaWindowState.Critical)]
    [InlineData(99, QuotaWindowState.Critical)]
    [InlineData(100, QuotaWindowState.Exhausted)]
    public void MapsAvailablePercentToWindowState(int used, QuotaWindowState expected)
    {
        var source = new RawQuotaWindow("5 小时", used, 300, null, false);

        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(expected, result.State);
    }

    [Fact]
    public void ConvertsDurationAndResetTimestamp()
    {
        var source = new RawQuotaWindow("5 小时", 40, 300, 1_800_000_000, false);

        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(TimeSpan.FromMinutes(300), result.WindowDuration);
        Assert.Equal(DateTimeOffset.FromUnixTimeSeconds(1_800_000_000), result.ResetsAt);
    }

    [Fact]
    public void UnlimitedWindowDoesNotPretendToHaveAPercentage()
    {
        var source = new RawQuotaWindow("5 小时", 80, 300, 1_800_000_000, true);

        var result = QuotaNormalizer.NormalizeWindow(source);

        Assert.Equal(QuotaWindowState.Unlimited, result.State);
        Assert.Null(result.AvailablePercent);
        Assert.Equal(TimeSpan.FromMinutes(300), result.WindowDuration);
        Assert.Equal(DateTimeOffset.FromUnixTimeSeconds(1_800_000_000), result.ResetsAt);
    }
}

File: macos/Packages/CodexQuotaKit/Tests/CodexQuotaKitTests/QuotaNormalizerTests.swift
import Foundation
import Testing
@testable import CodexQuotaKit

@Suite("额度归一化")
struct QuotaNormalizerTests {
    @Test("把已用额度转换为可用额度")
    func convertsUsedToAvailable() {
        let raw = RawQuotaSnapshot(
            primary: RawQuotaWindow(
                label: "5 小时",
                usedPercent: 68,
                windowDurationMinutes: 300,
                resetsAt: nil,
                isUnlimited: false),
            secondary: nil,
            receivedAt: Date(timeIntervalSince1970: 1_800_000_000))

        let state = QuotaNormalizer.normalize(raw)

        #expect(state.windows.map(\.availablePercent) == [32])
        #expect(state.connection == .live)
    }

    @Test("把异常百分比约束在零到一百")
    func clampsAvailablePercent() {
        #expect(QuotaNormalizer.availablePercent(from: -1) == 100)
        #expect(QuotaNormalizer.availablePercent(from: 0) == 100)
        #expect(QuotaNormalizer.availablePercent(from: 50) == 50)
        #expect(QuotaNormalizer.availablePercent(from: 80) == 20)
        #expect(QuotaNormalizer.availablePercent(from: 100) == 0)
        #expect(QuotaNormalizer.availablePercent(from: 101) == 0)
    }

    @Test("缺失百分比不伪造满额度")
    func missingPercentIsUnavailable() {
        let raw = RawQuotaSnapshot(
            primary: RawQuotaWindow(
                label: "主额度",
                usedPercent: nil,
                windowDurationMinutes: nil,
                resetsAt: nil,
                isUnlimited: false),
            secondary: nil,
            receivedAt: .now)

        let window = QuotaNormalizer.normalize(raw).windows[0]

        #expect(window.availablePercent == nil)
        #expect(window.state == .unavailable)
    }

    @Test("无限额度保持无限语义")
    func unlimitedStaysUnlimited() {
        let raw = RawQuotaSnapshot(
            primary: RawQuotaWindow(
                label: "5 小时",
                usedPercent: 99,
                windowDurationMinutes: 300,
                resetsAt: nil,
                isUnlimited: true),
            secondary: nil,
            receivedAt: .now)

        let window = QuotaNormalizer.normalize(raw).windows[0]

        #expect(window.availablePercent == nil)
        #expect(window.state == .unlimited)
    }

    @Test("颜色锚点从绿色过渡到红色")
    func colorAnchors() {
        #expect(QuotaColor.rgb(for: 100) == QuotaRGB(red: 0.57, green: 0.94, blue: 0.42))
        #expect(QuotaColor.rgb(for: 50) == QuotaRGB(red: 1.00, green: 0.77, blue: 0.36))
        #expect(QuotaColor.rgb(for: 20) == QuotaRGB(red: 1.00, green: 0.38, blue: 0.36))
        #expect(QuotaColor.rgb(for: 0) == QuotaRGB(red: 1.00, green: 0.38, blue: 0.36))
    }
}
